app/src/main/java/org/hwyl/sexytopo/control/graph/GraphView.java

```diff
@@ -1182,40 +1183,57 @@ private void drawSketch(Canvas canvas, Sketch sketch, int alpha) {
             return;
         }
 
+        drawPaint.setAlpha(alpha);
+
+        int lastColour = -1;
         for (PathDetail pathDetail : sketch.getPathDetails()) {
 
             if (!couldBeOnScreen(pathDetail)) {
                 continue;
             }
 
-            drawPaint.setColor(pathDetail.getColour().intValue);
-            drawPaint.setAlpha(alpha);
+            // Avoiding constantly updating the paint colour saves approx. 10% of sketch draw time.
+            // Ideally getPathDetails() would return the paths in colour order but HashSets
+            // are unordered collections
+            if (pathDetail.getColour().intValue != lastColour) {
+                lastColour = pathDetail.getColour().intValue;
+                drawPaint.setColor(lastColour);
+            }
 
             List<Coord2D> path = pathDetail.getPath();
 
-            Coord2D from = null;
             int lineIndex = 0;
+            float fromX = -1, fromY = -1;
             float[] lines = new float[path.size() * 4];
 
+            // This loop is the slowest part of the draw phase. Pulling out the calculations from
+            // within surveyCoordsToViewCoords saves a not insignificant amount of time by not
+            // constructing many thousands of Coord2D objects (approx. 10% of sketch draw time)
             for (Coord2D point : path) {
-                if (from == null) {
-                    from = surveyCoordsToViewCoords(point);
+                if (fromX == -1) {
+                    //from = surveyCoordsToViewCoords(point);
+                    fromX = (float)((point.x - viewpointOffset.x) * surveyToViewScale);
```

Comment:
I was thinking of having a method that calculated x and a separate one that calculated y and calling them from the coord translation method and here. Presumably the compiler would be smart enough to inline it. But we can go with this for now.

Reply:
Memory allocations are now our limiting factor. The profiler was showing that the main thread now spends (on that large survey you sent me) 50% of the time in `onDraw` and 50% idle which means the rest of the performance is affected either outside of user code of via memory allocations.

Reply:
Yep, definitely avoiding creating all those objects seems like the way to go.